APISpec.md

```diff
@@ -0,0 +1,508 @@
+```
+ID - Guid
+Token - String (JWT from auth0)
+User - User
+Expiration - Datetime
+```
+
+### Status
+```
+OPEN
+IN_PROGRESS
+COMPLETE
+```
+
+## Endpoints
+
+### Create New Todo
+---
+
+```
+POST /todos
+```
+
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body: 
+```
+{
+    "meta": {},
+    "data": {
+        "title": "Buy groceries",
+        "status": "OPEN|IN_PROGRESS|COMPLETE"
+        "dueDate": "2022-04-01T13:02:37.000Z"
+    }
+}
+```
+
+**Response** - `201 Created`
+
+Headers: 
+```
+"Location": "/todos/{id}"
+```
+
+**Error Codes**:
+
+`400 Bad Request` - Request payload invalid OR dueDate provided not a valid date
+
+`401 Unauthorized` - Request does not have valid authentication to create todo
+
+### Fetch all Todos
+---
+
+**Request**:
+``` 
+GET /todos
+```
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body: None
+
+**Response** - `200 OK`
+
+Body:
+```
+{
+    "meta": {
+        "todoCount": 2
+    }, 
+    "data": [
+        {
+            "id": "8a480c40-e4be-4b89-91c9-844de091a214",
+            "title": "A todo",
+            "status": "OPEN|IN_PROGRESS|COMPLETE",
+            "creationDate": "2022-03-18T13:03:03.000Z",
+            "dueDate": "2022-04-01T00:00:00.000Z"
+        },
+        {
+            ...
+        }
+    ]
+}
+```
+
+**Error Codes**:
+
+`404 Bad Request` - Request payload invalid OR status provided not a valid enum value OR invalid dates
+
+`401 Unauthorized` - Request does not have valid authentication to fetch todos
+
+### Update a Todo
+---
+
+**Request**:
+```
+PUT /todos/{id}
+```
+
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body:
+```
+{
+    "meta": {},
+    "data": {
+        "title": "My updated todo",
+        "status": "OPEN|IN_PROGRESS|COMPLETE",
+        "dueDate": "2022-04-02T00:00:00.000Z"
+    }
+}
+```
+
+**Response** - `204 No Content`
+
+**Error Codes**:
+
+`404 Not Found` - Todo with given ID does not exist in the API
+
+`400 Bad Request` - Request payload invalid OR status provided not a valid enum value
+
+`401 Unauthorized` - Request does not have valid authentication to update the todo
+
+### Delete a Todo
+---
+
+**Request**:
+```
+DELETE /todos/{id}
+```
+
+Headers:
+```
+"Authorization": Bearer <Token>
+```
+
+Body: None
+
+**Response** - `204 No Content`
+
+**Error Codes**:
+
+`404 Not Found` - Todo with given ID does not exist in the API
+
+`401 Unauthorized` - Request does not have valid authentication to delete todo
+
+### Create a User
+---
+
+**Request**:
+```
+POST /users
+```
+
+Body:
+```
+{
+    "meta": {},
+    "data": {
+        "username": "tom_jerry",
+        "password": "P@ssw0rd!"
+    }
+}
+```
+
+**Response** - `201 Created`
+
+Headers:
+```
+"Location": "/users/{id}"
+```
+
+**Error Codes**:
+
+`400 Bad Request` - Request payload invalid
+
+`409 Conflict` - User with username already exists in API
+Body:
+```
+{
+    "meta": {
+        "success": false
+    },
+    "data": {
+        "msg": "username already exists"
```

Comment:
a status message in our spec would probably go into the `meta` object instead as the message isn't technically the payload

Reply:
updated